dipy/denoise/pca_noise_estimate.pyx

```diff
@@ -0,0 +1,150 @@
+"""
+================================
+PCA Based Local Noise Estimation
+================================
+
+"""
+
+import numpy as np
+import nibabel as nib
+import scipy.special as sps
+from scipy import ndimage
+cimport cython
+cimport numpy as cnp
+
+# Try to get the SVD through direct API to lapack:
+try:
+    from scipy.linalg.lapack import sgesvd as svd
+    svd_args = [1, 0]
+# If you have an older version of scipy, we fall back
+# on the standard scipy SVD API:
+except ImportError:
+    from scipy.linalg import svd
+    svd_args = [False]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def pca_noise_estimate(data, gtab, patch_radius=1, correct_bias=True, smooth=2):
+    """ PCA based local noise estimation.
+
+    Parameters
+    ----------
+    data: 4D array
+        the input dMRI data.
+
+    gtab: gradient table object
+      gradient information for the data gives us the bvals and bvecs of
+      diffusion data, which is needed here to select between the noise
+      estimation methods.
+    patch_radius : int
+        The radius of the local patch to be taken around each voxel (in
+        voxels). Default: 1 (estimate noise in blocks of 3x3x3 voxels).
+    correct_bias : bool
+      Whether to correct for bias due to Rician noise. This is an implementation
+      of equation 8 in [1]_.
+
+    smooth : int
+      Radius of a Gaussian smoothing filter to apply to the noise estimate
+      before returning. Default: no smoothing.
+
+    Returns
+    -------
+    sigma_corr: 3D array
+        The local noise standard deviation estimate.
+
+    References
+    ----------
+    .. [1] Manjon JV, Coupe P, Concha L, Buades A, Collins DL "Diffusion
+           Weighted Image Denoising Using Overcomplete Local PCA". PLoS ONE
+           8(9): e73021. doi:10.1371/journal.pone.0073021.
+    """
+    # first identify the number of the b0 images
+    K = gtab.b0s_mask[gtab.b0s_mask].size
+
+    if(K > 1):
+        # If multiple b0 values then
+        data0 = data[..., gtab.b0s_mask]
+
+    else:
+        # if only one b0 value then
+        # SIBE Noise Estimate
+        data0 = data[..., ~gtab.b0s_mask]
+        # MUBE Noise Estimate
+
+    cdef:
+        cnp.npy_intp n0 = data0.shape[0]
+        cnp.npy_intp n1 = data0.shape[1]
+        cnp.npy_intp n2 = data0.shape[2]
+        cnp.npy_intp n3 = data0.shape[3]
+        cnp.npy_intp nsamples = n0 * n1 * n2
+        cnp.npy_intp i, j, k, i0, j0, k0, l0
+        cnp.npy_intp pr = patch_radius
+        cnp.npy_intp  patch_size = 2 * pr + 1
+        double norm = patch_size ** 3
+        double sum_reg, temp1
+        double[:, :, :] I = np.zeros((n0, n1, n2))
+        double[:, :, :] count = np.zeros((n0, n1, n2))
+        double[:, :, :] mean = np.zeros((n0, n1, n2))
+        double[:, :, :] sigma_sq = np.zeros((n0, n1, n2))
+        double[:, :, :, :] data0temp = data0
+
+    X = data0.reshape(nsamples, n3)
```

Comment:
I think this is incorrect!
For Mube, X should have as many columns as image voxels and as many rows as images.
For Sibe, X should be reconstructed from the patches. Note: number of voxels should always the larger than number of images.